// @zk-ratemyprof/api/src/lib/zk-keys/verification-key.ts
/**
 * Embedded Verification Key for BRACU Email Verifier Circuit
 * 
 * Generated from trusted setup using Hermez Powers of Tau (2^20)
 * Circuit: bracu_verifier_simple.circom
 * Constraints: 704,052 non-linear
 * 
 * DO NOT MODIFY - This file is auto-generated
 */

export const BRACU_VERIFICATION_KEY = {
    protocol: "groth16",
    curve: "bn128",
    nPublic: 19,
    vk_alpha_1: [
        "20491192805390485299153009773594534940189261866228447918068658471970481763042",
        "9383485363053290200918347156157836566562967994039712273449902621266178545958",
        "1"
    ],
    vk_beta_2: [
        [
            "6375614351688725206403948262868962793625744043794305715222011528459656738731",
            "4252822878758300859123897981450591353533073413197771768651442665752259397132"
        ],
        [
            "10505242626370262277552901082094356697409835680220590971873171140371331206856",
            "21847035105528745403288232691147584728191162732299865338377159692350059136679"
        ],
        ["1", "0"]
    ],
    vk_gamma_2: [
        [
            "10857046999023057135944570762232829481370756359578518086990519993285655852781",
            "11559732032986387107991004021392285783925812861821192530917403151452391805634"
        ],
        [
            "8495653923123431417604973247489272438418190587263600148770280649306958101930",
            "4082367875863433681332203403145435568316851327593401208105741076214120093531"
        ],
        ["1", "0"]
    ],
    vk_delta_2: [
        [
            "10857046999023057135944570762232829481370756359578518086990519993285655852781",
            "11559732032986387107991004021392285783925812861821192530917403151452391805634"
        ],
        [
            "8495653923123431417604973247489272438418190587263600148770280649306958101930",
            "4082367875863433681332203403145435568316851327593401208105741076214120093531"
        ],
        ["1", "0"]
    ],
    vk_alphabeta_12: [
        [
            [
                "2029413683389138792403550203267699914886160938906632433982220835551125967885",
                "21072700047562757817161031222997517981543347628379360635925549008442030252106"
            ],
            [
                "5940354580057074848093997050200682056184807770593307860589430076672439820312",
                "12156638873931618554171829126792193045421052652279363021382169897324752428276"
            ],
            [
                "7898200236362823042373859371574133993780991612861777490112507062703164551277",
                "7074218545237549455313236346927434013100842096812539264420499035217050630853"
            ]
        ],
        [
            [
                "7077479683546002997211712695946002074877511277312570035766170199895071832130",
                "10093483419865920389913245021038182291233451549023025229112148274109565435465"
            ],
            [
                "4595479056700221319381530156280926371456704509942304414423590385166031118820",
                "19831328484489333784475432780421641293929726139240675179672856274388269393268"
            ],
            [
                "11934129596455521040620786944827826205713621633706285934057045369193958244500",
                "8037395052364110730298837004334506829870972346962140206007064471173334027475"
            ]
        ]
    ],
    IC: [
        ["21127408134299559613135026370401224275537496975928807890750800347357811623991", "19062931075813782350527504054361417291592689298800993830807618226115259474843", "1"],
        ["10390493343640539705501594076499336051171879402946173485635154466478385011526", "738675736659766073234697520640791751820794622256499884522817283345878594025", "1"],
        ["14896750507971467644230409123148810812028937268253566565190366369996100546224", "10488853571463064379771668348137944458154382456904031539762071656854900782814", "1"],
        ["9851490995319512850616172543600811642419611390669202207432624075385577602600", "2098827492097254194465746953271363880930254555293413687186831031732516816739", "1"],
        ["21363566272586488019433623800100972984629862072787807910273124175566454171187", "13848386549642400231899322131541969513416069080571984069868680441022876593774", "1"],
        ["12727680257997711543279236257964372765704103508325448173658307080322962139036", "12902805712857598165391192019273914805595296059305835812258521232093426245405", "1"],
        ["20434861778336828117856472697575095038904232251231811894734323552045740881107", "12976008927451843687104998035149584504464929432538243196870312793745302181357", "1"],
        ["11136800476920443479911252541644484625751538593347352697595671688915024891506", "12276373811143357040826268074841496545766043813953137106326778701899564857327", "1"],
        ["11650361925110742539280919353421005636534351509820501380787677996619991631759", "14947722491410298368217135294675250587696078836310217196864195458071938730571", "1"],
        ["21363089626067500930639351420371605716589493342286140178579361481910285502283", "6263088055614416615371291940896908728717195900322136212531942966432113131040", "1"],
        ["18794808781837254168530675799509685514357643647899337332394829012759657061961", "6210699017105605136493293569567941985050226752352019066801180032081665123176", "1"],
        ["20091454418936136696839877401587201164476033016906759644717428969969673291486", "10565393682239481751665751844348776111673314838988971831978844895467438480857", "1"],
        ["9142464267039730859911161716751758729997804331731581314352140734937447765199", "4112067946719095130300507454350385517268701946529513831737518351156712253994", "1"],
        ["17340210020293968128095487420583384862377347162853652576627650516145908308761", "3870367128056258817866937850653282635807149643524817934151048599991249652834", "1"],
        ["10776166950334912822393072782157740350372852819365866396473305254366865068567", "14177322005927637565513944467944417824279967941324288607966376295888818525335", "1"],
        ["11729497514106603091641688836826788833313168610397901322242528896678678949405", "9356000305430787866787255962855488245926280241219774156825686923304426724583", "1"],
        ["12821894070933272700207217785476540840150642020944294200486812595364672246533", "4220321023252103419660364438631943695066530777802585339504214364774753624519", "1"],
        ["11923445261229334852529830956019127214504490520786449922229198113900059409860", "10271612413042821584909057177916897624404761015453250480850953836719218887585", "1"],
        ["3228040021445395523143141182303487526293947270685005497215527086287133446473", "18623163585425923043116393845866700065886116284849958601273317793790133524069", "1"],
        ["2133746681042987876647544372071538871522681027184263023926635352047439746195", "140013327424762733294947022367560453228045910956448279344991430154474156680", "1"]
    ]
} as const;

export type BracuVerificationKey = typeof BRACU_VERIFICATION_KEY;
